### PR TITLE
Make /sync return heroes if room name or canonical alias are empty

### PR DESCRIPTION
**Pull request by dnaf**
This closes issue #4194

### CODE REVIEW

**Comment by richvdh** [comment]
Hi, and thanks very much for your work on this. However, I think you've stumbled into a hornet's nest, which is the lack of specification for this feature (see https://github.com/matrix-org/matrix-doc/issues/1868).

I feel quite strongly that if this is a thing that's important enough to change, then it ought to be specced and tested to make sure that clients agree on it being the right thing to do, and future server implementations do the same thing.

In other words: I don't think we can progress this until we have a spec for `m.heroes`, and sytests to validate that it is followed.

**Comment by richvdh** [comment]
[a bunch of the CI has failed here due to a problem which is fixed in latest `develop`.]

**Comment by babolivier** [comment]
Given we're looking at some lazy loading bugs in the context of the upcoming Synapse 1.0, I've done some updates on this one in the hope to get it merged. What I've done is merging develop in to unwedge the CI, and removing `[event].content` from the condition since the spec forbids the content of `m.room.name` and `m.room.canonical_alias` events from being either null or an empty dict (and `[event].content.get` covers the latter anyway).

File: synapse/handlers/sync.py [hunk 1]
```diff
@@ -583,19 +583,18 @@ def compute_summary(self, room_id, sync_config, batch, state, now_token):
...
-            if name and name.content:
+            if name and name.content.get("name", ""):
                 defer.returnValue(summary)
 
         if canonical_alias_id:
             canonical_alias = yield self.store.get_event(
                 canonical_alias_id, allow_none=True,
             )
-            if canonical_alias and canonical_alias.content:
+            if canonical_alias and canonical_alias.content.get("alias", ""):
```
Comment (erikjohnston) [inline synapse/handlers/sync.py]: You can' just write this as `canonical_alias.content.get("alias")` since it'll default to `None` which is falsey

Comment (babolivier) [reply synapse/handlers/sync.py]: True.

**Comment by babolivier** [comment]
One more checked off the list, thanks for your work @dnaf! :slightly_smiling_face:
